proj/hps_accel/gateware/gen2/macc.py

```diff
@@ -0,0 +1,162 @@
+        The value of input_b, delayed by one cycle.
+    input_first: Signal(1), in
+        Indicates that this is the first value passed, and the
+        accumulator should be assumed zero.
+    output_first: Signal(1), out
+        The value of input_first, delayed by one cycle
+    output_last: Signal(1), out
+        The value of output_last, delayed by one cycle
+    output_accumulator: accumulator_shape, out
+        Holds the output value. Since there is no flow control, this
+        value must be read before it is updated again.
+    """
+
+    # The latency from input_received to accumulator being updated
+    # Moving the accumulator to accumulator_output takes an additional
+    # cycle.
+    PIPELINE_CYCLES = 2
+
+    def __init__(self, n, a_shape, b_shape, accumulator_shape):
+        self._n = n
+        self._a_shape = a_shape
+        self._b_shape = b_shape
+        self._accumulator_shape = accumulator_shape
+
+        self.input_a = Signal(unsigned(n * a_shape.width))
+        self.output_a = Signal.like(self.input_a)
+        self.input_b = Signal(unsigned(n * b_shape.width))
+        self.output_b = Signal.like(self.input_b)
+
+        self.input_first = Signal()
+        self.output_first = Signal()
+        self.input_last = Signal()
+        self.output_last = Signal()
+
+        self.output_accumulator = Signal(accumulator_shape)
+
+    def _connect_passthrough(self, m):
+        """Connects the block pass through signals."""
+        m.d.sync += [
+            self.output_a.eq(self.input_a),
+            self.output_b.eq(self.input_b),
+            self.output_first.eq(self.input_first),
+            self.output_last.eq(self.input_last),
+        ]
+
+    @staticmethod
+    def _delay(m, sig, cycles):
+        """Creates simple shift register that delays a bit signal."""
+        shift_register = Signal(cycles)
+        m.d.sync += [
+            shift_register[1:].eq(shift_register[0:]),
+            shift_register[0].eq(sig)
+        ]
+        return shift_register
+
+    def elab(self, m):
+        self._connect_passthrough(m)
+
+        # Pipeline cycle 0: calculate products
+        products = []
+        for i in range(self._n):
+            a_bits = self.input_a.word_select(i, self._a_shape.width)
+            b_bits = self.input_b.word_select(i, self._b_shape.width)
+            a = Signal(self._a_shape)
+            b = Signal(self._b_shape)
+            m.d.comb += [
+                a.eq(a_bits),
+                b.eq(b_bits),
+            ]
+            ab = Signal.like(a * b)
+            m.d.sync += ab.eq(a * b)
+            products.append(ab)
+
+        # Pipeline cycle 1: accumulate
+        product_sum = Signal.like(tree_sum(products))
+        m.d.comb += product_sum.eq(tree_sum(products))
+        first_delayed = self._delay(m, self.input_first, 1)
+        accumulator = Signal(self._accumulator_shape)
+        base = Mux(first_delayed[0], 0, accumulator)
```

Comment:
Is [0] the input end of the delay register, or the output end?

Reply:
Yeah, I'm not 100% happy with the API. Will probably make it less noisy in future.

Reply:
I made a better API in the next PR.